src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs

```diff
@@ -101,7 +104,46 @@ public static IWebHostBuilder UseServer(this IWebHostBuilder hostBuilder, string
                 throw new ArgumentNullException(nameof(assemblyName));
             }
 
-            return hostBuilder.UseSetting(WebHostDefaults.ServerKey, assemblyName);
+            var assembly = Assembly.Load(new AssemblyName(assemblyName));
+            if (assembly == null)
+            {
+                throw new InvalidOperationException(string.Format("The assembly {0} failed to load.", assemblyName));
+            }
+
+            var serverTypeInfo = assembly.DefinedTypes.Where(
+                t => t.ImplementedInterfaces.FirstOrDefault(interf => interf.Equals(typeof(IServerFactory))) != null)
+                .FirstOrDefault();
+
+            if (serverTypeInfo == null)
+            {
+                throw new InvalidOperationException($"No server type found that implements IServerFactory in assembly: {assemblyName}.");
+            }
+
+            return hostBuilder.ConfigureServices(services =>
+            {
+                services.AddSingleton(typeof(IServerFactory), serverTypeInfo.AsType());
```

Comment:
Can't be transient. See the other place where we register this. Unless we remove the instance overloads it's bad to use different lifetimes for the same interface.